model/src/main/kotlin/DependencyNavigator.kt

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (C) 2021 Bosch.IO GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.model
+
+/**
+ * Alias for a function that is used while navigating through the dependency graph to determine which dependencies
+ * should be followed or skipped. The function is passed the current dependency node; it returns a flag whether this
+ * dependency is matched or not.
+ */
+typealias DependencyMatcher = (DependencyNode) -> Boolean
+
+/**
+ * An interface allowing the navigation through the dependency information contained in an [OrtResult], independent on
+ * the concrete storage representation.
+ *
+ * The dependencies detected by the analyzer are represented in an [OrtResult] in an optimized form which is not
+ * accessible easily. To simplify dealing with dependencies, this interface is introduced. It offers functionality
+ * tailored towards the typical use cases required by the single ORT components.
+ */
+interface DependencyNavigator {
+    companion object {
+        /**
+         * A pre-defined [DependencyMatcher] that matches all dependencies. It can be used to traverse the whole
+         * dependency graph.
+         */
+        val MATCH_ALL: DependencyMatcher = { true }
+    }
+
+    /**
+     * Return a set with the names of all the scopes defined for the given [project].
+     */
+    fun scopeNames(project: Project): Set<String>
```

Comment:
The function could have a default implementation `= scopeNames(project.id)`. Same for the functions below. Is the duplication of the function for `project` and `projectId` really required?

Reply:
This duplication bothers me, too. I took another look at the call sites, and it seems that always a full `Project` is used. So I will remove the variants that work on a project ID; this will get rid off a bunch of functions, and also the test class gets much leaner.